src/escrows/LPCurveYearnV2Escrow.sol

```diff
@@ -73,31 +90,43 @@ contract DolaFraxBPEscrow {
     function pay(address recipient, uint amount) public {
         if (msg.sender != market) revert OnlyMarket();
         uint256 tokenBal = token.balanceOf(address(this));
-        uint256 missingAmount = amount > tokenBal ? amount - tokenBal : 0;
-        // If there are enough tokens in the escrow, transfer the amount
+
         if (tokenBal >= amount) {
             token.safeTransfer(recipient, amount);
-        } else if (stakedBalance >= missingAmount) {
-            // If there are enough staked tokens in convex, withdraw the amount from convex
-            stakedBalance -= missingAmount;
-            rewardPool.withdrawAndUnwrap(missingAmount, false);
-            token.safeTransfer(recipient, amount);
-        } else if (
-            YearnVaultV2Helper.collateralToAsset(
-                yearn,
-                yearn.balanceOf(address(this))
-            ) >= missingAmount
-        ) {
-            // If there are enough tokens in Yearn, withdraw the amount from Yearn
-            uint256 withdrawAmount = YearnVaultV2Helper.assetToCollateral(
+            return;
+        }
+
+        uint256 missingAmount = amount - tokenBal;
+        uint256 convexBalance = rewardPool.balanceOf(address(this));
+        if (convexBalance > 0 && missingAmount > 0) {
+            uint256 withdrawAmount = convexBalance > missingAmount
+                ? missingAmount
+                : convexBalance;
+            missingAmount -= withdrawAmount;
+            rewardPool.withdrawAndUnwrap(withdrawAmount, false);
+        }
+
+        uint yearnBal = yearn.balanceOf(address(this));
+        if (yearnBal > 0 && missingAmount > 0) {
+            uint256 maxWithdraw = YearnVaultV2Helper.collateralToAsset(
                 yearn,
-                missingAmount
+                yearnBal
             );
+            uint256 withdrawAmount = maxWithdraw > missingAmount
+                ? missingAmount
+                : maxWithdraw;
+
+            uint256 collateralAmount;
+            if (withdrawAmount == maxWithdraw) collateralAmount = yearnBal;
+            else
+                collateralAmount = YearnVaultV2Helper.assetToCollateral(
+                    yearn,
+                    withdrawAmount + weiDelta
+                );
             // Withdraw from Yearn
-            uint256 assetAmount = yearn.withdraw(withdrawAmount, address(this));
-            // Transfer the amount to the recipient
-            token.safeTransfer(recipient, assetAmount + tokenBal);
+            yearn.withdraw(collateralAmount, address(this));
```

Comment:
Suggestion: in an extremely unlikely scenario, there can be a loss on a yearn vault. If a loss exceeds 0.01% (default in the vault), the escrow's position will be stuck.

i would suggest replacing your current interface with one that includes the optional `maxLoss` parameter. and allow users to adjust this value. 
Reference on vault interface is here...

https://github.com/yearn/yearn-vaults/blob/97ca1b2e4fcf20f4be0ff456dabd020bfeb6697b/contracts/Vault.vy#L1033

Reply:
Same for line 247.